protocol/nats_jetstream/v3/options.go

```diff
@@ -0,0 +1,123 @@
+/*
+ Copyright 2024 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"errors"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+var (
+	ErrNoConnection              = errors.New("URL or nats connection must be given")
+	ErrNoFilterSubjects          = errors.New("no filter subjects were given")
+	ErrMoreThanOneStream         = errors.New("more than one stream for given filter subjects")
+	ErrNoConsumerConfig          = errors.New("no consumer config was given")
+	ErrMoreThanOneConsumerConfig = errors.New("more than one consumer config given")
+)
+
+// ProtocolOption is the function signature required to be considered an nats.ProtocolOption.
+type ProtocolOption func(*Protocol) error
+
+// WithURL configures the Sender and/or Receiver
+func WithURL(url string, natsOpts ...nats.Option) ProtocolOption {
```

Comment:
There's quite a lot of logic in this option. I thought URL was always required? If so, better to make it a constructor parameter and configure these fields after all options are applied.

Reply:
You can use one of two options:
1) URL +nats.Options  // if you do not have a connection yet
2) *nats.Conn // if you already have a connection.

This requires either:
1) two constructors
   a) context.Context, URL , nats.Options,   ...ProtocolOption
   b) context.Context, *nats.Conn,   ...ProtocolOption
2) Options that makes the two cases "mutually exclusive"

Reply:
Simplified more.  One constructor.

Reply:
The `natsOpts ...nats.Option` is critical.  In fact, we could not use this option without the nats.Option argument.  If youe NATS server needs any sort of TLS or authentication with JWTs, these options must be exposed.

Reply:
see:
https://docs.nats.io/using-nats/developer/connecting/creds

or any of the other authentication methods of NATS.

Reply:
I see what you mean. Well, it does break the functional options pattern a bit IMHO. I'd expect another option to configure nats options in fact e.g., `WithNatsOptions`. I can see this being useful for other scenarios so can we add this option?

Also this is a bit confusing

> If WithConnection is also used, WithURL will be ignored.

Because ultimately if you specify both options our validation would fail. Can you please change the comment to say that WithURL cannot be used with WithConnection?

Reply:
done